modules/3d/src/rgbd/odometry_functions.cpp

```diff
@@ -0,0 +1,1645 @@
+                // and gather sum
+                int pos = 0;
+                for (int i = 0; i < 6; i++)
+                {
+                    for (int j = i; j < 7; j++)
+                    {
+                        upperTriangle[pos++] += ab[i] * ab[j];
+                    }
+                }
+            }
+        }
+
+        ABtype sumAB = ABtype::zeros();
+        int pos = 0;
+        for (int i = 0; i < 6; i++)
+        {
+            for (int j = i; j < 7; j++)
+            {
+                sumAB(i, j) = upperTriangle[pos++];
+            }
+        }
+#endif
+        AutoLock al(mtx);
+        globalSumAb += sumAB;
+    }
+
+    ABtype& globalSumAb;
+    Mutex& mtx;
+    const Points& oldPts;
+    const Normals& oldNrm;
+    const Points& newPts;
+    const Normals& newNrm;
+    Affine3f pose;
+    const Intr::Projector proj;
+    float sqDistanceThresh;
+    float minCos;
+};
+
+void calcICPLsmMatricesFast(Matx33f cameraMatrix, const Mat& oldPts, const Mat& oldNrm, const Mat& newPts, const Mat& newNrm,
+    cv::Affine3f pose, int lvls, int level, float maxDepthDiff, float angleThreshold, cv::Matx66f& A, cv::Vec6f& b)
+{
+    CV_Assert(oldPts.size() == oldNrm.size());
+    CV_Assert(newPts.size() == newNrm.size());
+    CV_Assert(lvls > 0);
+
+#ifdef HAVE_OPENCL_
+    AccessFlag af = AccessFlag::ACCESS_READ;
+    calcICPLsmMatricesFast_ocl(cameraMatrix,
+        oldPts.getUMat(af), oldNrm.getUMat(af),
+        newPts.getUMat(af), newNrm.getUMat(af),
+        pose, lvls, level, maxDepthDiff, angleThreshold,
+        A, b);
+    return;
+#endif
+
+    ABtype sumAB = ABtype::zeros();
+    Mutex mutex;
+    const Points  op(oldPts), np(newPts);
+    const Normals on(oldNrm),  nn(newNrm);
+
+
+    Intr intrinsics(cameraMatrix);
+    GetAbInvoker invoker(sumAB, mutex, op, on, np, nn, pose,
+        intrinsics.scale(level).makeProjector(),
+        maxDepthDiff * maxDepthDiff, std::cos(angleThreshold));
+    Range range(0, newPts.rows);
+    const int nstripes = -1;
+    parallel_for_(range, invoker, nstripes);
+
+    // splitting AB matrix to A and b
+    for (int i = 0; i < 6; i++)
+    {
+        // augment lower triangle of A by symmetry
+        for (int j = i; j < 6; j++)
+        {
+            A(i, j) = A(j, i) = sumAB(i, j);
+        }
+
+        b(i) = sumAB(i, 6);
+    }
+}
+
+#ifdef HAVE_OPENCL
+
+void calcICPLsmMatricesFast_ocl(Matx33f cameraMatrix, const UMat& oldPts, const UMat& oldNrm, const UMat& newPts, const UMat& newNrm,
+    cv::Affine3f pose, int lvls, int level, float maxDepthDiff, float angleThreshold, cv::Matx66f& A, cv::Vec6f& b)
+{
+    CV_TRACE_FUNCTION();
+
+    Size oldSize = oldPts.size(), newSize = newPts.size();
+    CV_Assert(oldSize == oldNrm.size());
+    CV_Assert(newSize == newNrm.size());
+
+    // calculate 1x7 vector ab to produce b and upper triangle of A:
+    // [A|b] = ab*(ab^t)
+    // and then reduce it across work groups
+
+    std::vector<UMat> groupedSumBuffers(lvls);
```

Comment:
Why do you need `lvls` here? Why do you need a whole vector of UMats here? Isn't one UMat is enough?

Reply:
done